### PR TITLE
fix: keep the odd XP when a student's total is split between courses (#2490)

### LINKED ISSUE #2474: Should a badge granted between terms follow the student into the term they join?
Follow-up from #2473, which settled #2441 and #2413.

### Context

Work handed in by someone registered in no semester is now stamped with no semester, and joining a course brings the quests they had **in progress** into that semester (`QuestSubmission.objects.adopt_unstamped_in_progress()`, called from a `post_save` receiver on `CourseStudent` in `src/courses/models.py`). Without that, an in-progress quest would be stranded: out of the student's in-progress list, which is their new semester's, and out of their available list, which drops a quest they already have a submission of.

Badges were deliberately left out of that. `BadgeAssertion.objects.create_assertion()` stamps the same way (an assertion granted to someone in no semester belongs to none), but an assertion is an award already made rather than work in progress, so nothing is stranded: the student can see it on their badges page either way. Its XP simply doesn't count toward the term they then join.

### The question

Is that the behaviour a teacher expects? Two readings:

* **Leave it (current).** A badge granted between terms recognises what the student did between terms, so it belongs to no term. XP resets each semester for everyone, and this is the same rule.
* **Adopt it.** A teacher granting a badge to a student who is about to join a course probably means it to count in that course, and losing it the moment they register is surprising.

The same question applies to *completed* quest submissions from between terms, which #2473 also left unstamped for the same reason. Whatever is decided should probably apply to both, since a student can't tell why one carried and the other didn't.

If adoption is wanted, the mechanism already exists: extend the `CourseStudent` receiver to re-stamp the user's unstamped assertions (and completed submissions) with the semester they joined. The hazard to think about is how far back to reach: `semester` is `on_delete=SET_NULL` on both models, so deleting an old semester leaves ancient rows unstamped, and a blanket adoption would sweep those into the new term.

- Claude Code (`bytedeck - semester workflow redesign`)

### LINKED ISSUE #2490: `final_xp` silently truncates XP when shared XP cannot be divided evenly across a multi-course student's courses
### What happens

When a semester is archived, a student enrolled in more than one course can permanently **lose XP from their final record** — silently, with no error and no warning.

`CourseStudent.xp()` returns a float whenever the student has unassigned (shared) XP:

```python
# src/courses/models.py
return assigned_here + self.xp_adjustment + shared / len(current_course_ids)
```

`calc_semester_grades()` assigns this value directly to `final_xp`, which is a `PositiveIntegerField`. Django's `IntegerField.get_prep_value` applies `int(value)`, which truncates toward zero on a per-row basis.

### Example

Consider a student enrolled in 2 courses with 5 XP of unassigned work:

- each registration's `xp()` returns `2.5`
- stored `final_xp` → `2` and `2`
- recorded total: **4 ≠ 5**

With 3 courses and 10 shared XP: each share is `3.33`, truncated to `3`; 3 × 3 = 9 ≠ 10. This loss is deterministic (the remainder is always dropped), not an occasional floating-point artifact.

The invariant that should hold: **the sum of all registrations' `final_xp` for shared XP should equal the shared XP total.**

### How to reproduce

1. Enable *Students choose which course their XP counts toward* (enabled by default)
2. Register one student in two courses
3. Have the student complete quests that grant XP **without** selecting a course (the default "split evenly" path), with a total shared XP that is not divisible by 2
4. Archive the semester
5. Both `final_xp` values are truncated, and the remainder is permanently lost from the record

Existing tests for `calc_semester_grades` use evenly divisible amounts (60, 500), so this edge case passes CI today.

### Suggested fix

Perform the integer allocation once at the group level, rather than truncating on a per-row basis — e.g., `divmod(shared, n_courses)` and distribute the remainder to the first `r` registrations under a stable ordering (largest-remainder style). This preserves `final_xp` as an integer field while ensuring that the totals are conserved.

Alternatively, `xp()` could return exact shares and the field could be changed to a `DecimalField`, but that would alter the field's semantics for every consumer. The integer allocation approach appears to be the smaller and safer change.

A regression test worth adding would assert that the sum of `final_xp` across a multi-course student's registrations equals the total XP earned, using a non-divisible case.

### Context

This was found during a systematic review of #2456 and the semester-archiving code path. I would be happy to provide more detail or a patch if that would be helpful.

### PR DESCRIPTION
**Pull request by tylerecouture**
### What?

Closes #2490. A student in more than one course keeps every XP they earned when their semester is archived.

* New `whole_xp_shares()` in `src/courses/models.py` rounds a student's per-course shares once for the student as a whole, by the largest-remainder method, so the shares are whole XP that still add up to the student's total. `split_xp_between_registrations()` finishes with it, which puts it on every path that divides XP: the marks page, the per-course ranks in the navbar, the XP progress chart, and `calc_semester_grades()`.
* Also closes #2474, with a regression test rather than a code change: a badge granted between semesters does not follow the student into the term they join. That is what the code already does and nothing pinned it down, so widening the adoption receiver from quests to badges would not have failed anything.

No migration, no model change.

### Why?

Reported by @doudou868. `CourseStudent.final_xp` is a `PositiveIntegerField`, and the share written into it is often not a whole number, so Django's `int()` rounded each row down on its own and the rest of the student's work vanished from their permanent record with no error and no warning.

Two courses sharing 5 XP archived as 2 and 2. Three sharing 10 archived as 3, 3 and 3. The loss is deterministic, not an occasional floating-point artifact, and the existing tests missed it because they used amounts that divide evenly (60, 500).

Two things make a share fractional, and the issue names the first:

1. XP the student did not assign to a particular course is shared evenly between the courses they hold.
2. A repeatable quest taken past its `max_xp` gives each course the capped total in proportion to what it was assigned uncapped (#2440). That is a fraction even when nothing is being shared: a quest capped at 15 with 10 XP against each of two courses gives 7.5 and 7.5, which archived as 7 and 7.

Rounding at the point of writing `final_xp` would have fixed the record and left the deck disagreeing with itself, because the same fractions are what the student is shown all term. Rounding the division itself fixes both at once, and means the number a student sees against a course during the term is the number archived against it for good.

### How?

The shares are worked out exactly as before, then handed to `whole_xp_shares()`, which rounds each one down and gives the odd XP back a point at a time to the shares that came closest to earning it:

```python
whole = [math.floor(share) for share in exact_shares]
odd_xp = round(sum(exact_shares)) - sum(whole)

owed_most_first = sorted(
    range(len(exact_shares)),
    key=lambda index: (whole[index] - exact_shares[index], registrations[index].pk),
)
for index in owed_most_first[:odd_xp]:
    whole[index] += 1
```

That leaves every course within 1 XP of its exact share and the shares adding up to the student's total. `math.floor` rather than `int` because a share can be negative (a large enough negative `xp_adjustment`), and truncating toward zero there would hand out XP nobody earned.

Ties break on `pk`, the registration made first, rather than on the order the caller passed. That matters because the live path (`current_courses()`) and the archive path (`get_semester()`) can order registrations differently: keying on the row itself makes the division a property of the student rather than of the question asked, so the marks page and the archived record cannot disagree about which course got the odd point. The returned list keeps the order it was given, which `split_xp_between_registrations()` relies on to zip the shares back onto their registrations; both that and the tie-break rule are stated in the docstring.

### Testing?

`python src/manage.py test src` (2628 tests, all passing), `ruff check src` clean, `manage.py check` clean, `makemigrations --check --dry-run` clean. `courses/models.py` is at 99% line coverage with the three misses pre-existing and outside this diff; `whole_xp_shares()` is fully covered, and Codecov reports every modified line covered with project coverage unchanged at 99.93%.

Five new tests. The four for #2490 were each run against the code without the fix and fail there, with exactly the losses the issue describes:

| test | without the fix |
|---|---|
| `test_calc_semester_grades__records_the_odd_xp_rather_than_dropping_it` | `4 != 5` |
| `test_calc_semester_grades__hands_the_odd_xp_out_a_point_at_a_time` | `9 != 10` |
| `test_calc_semester_grades__records_the_xp_the_student_was_shown_all_term` | `[3, 3] != [3.5, 3.5]` |
| `test_xp_across__gives_whole_xp_when_a_capped_quest_is_split_between_courses` | `[7.5, 7.5] != [7, 8]` |

The fifth, `test_create_assertion__a_badge_granted_between_semesters_does_not_follow_the_student`, is a lock-in test for the #2474 decision rather than a bug fix, so it passes both with and without this diff by design.

Walked through the running dev deck as well: a student on 175 XP in two courses, archived before and after the change (screenshots below).

### Screenshots (if front end is affected)

All captured from the running dev deck at `http://hackerspace.localhost:8000`, logged in as a student registered in Mathematics 10 and Digital Art 11 with 175 XP, none of it assigned to a particular course.

**The permanent record**, on their profile after the semester is archived. This is the bug: they earned 175 and the record says 174.

Before, 87 + 87:

archived record before

After, 87 + 88:

archived record after

**The marks page**, during the term. The same fractions that were being lost at archiving were also what the student was reading all term.

Before, both courses on 87.5 XP and the same 67.3%:

marks page before

After, 87 and 88, and the two courses no longer report an identical mark:

marks page after

### Anything Else?

Every XP figure a student sees against a course is now a whole number, which is a visible change beyond the archived record: the marks page, the per-course ranks in the navbar and the XP progress chart all showed `87.5` and now show `87` or `88`. That reads as a correction rather than a regression (XP is a whole-point currency everywhere else in the app, and no other screen has ever shown a fraction of one), but it is worth knowing it is deliberate.

The alternative the issue raises, making `final_xp` a `DecimalField`, is not taken: it would change the field's meaning for every consumer and leave the student's own view of their XP fractional, which is the part they actually read.

One pre-existing oddity is left alone rather than widened into this PR: if a student somehow holds two registrations for the *same* course in one semester (`unique_together` allows it when the grades differ), each is credited with that course's whole assigned total, so the shares can add up to more than the student earned. `whole_xp_shares()` rounds to the exact shares' own total rather than to `xp_cached`, so it stays honest about whatever it is given instead of quietly absorbing that.

CodeRabbit's review raised two documentation points, both taken in `2fd55ffa`: the tie-break and order contract described above, and four docstrings (the function's own plus three of the new tests) that described the rounding as a change from the previous behaviour rather than as the invariant they hold, which stops meaning anything once this squash-merges. `develop` is merged in, so the branch is current.

### Review request
@tylerecouture

- Claude Code (`bytedeck - semester workflow redesign`)

## Summary by CodeRabbit

* **Bug Fixes**
  * XP earned across multiple course registrations is now allocated in whole numbers while preserving the correct total.
  * Remaining XP is distributed consistently when allocations cannot be divided evenly.
  * Capped quest XP now displays and distributes accurate whole-number shares.
  * Badge assertions created between semesters remain unstamped after later enrollment.

* **Tests**
  * Added regression coverage for XP allocation and cross-semester badge behavior.

### CODE REVIEW

**Comment by tylerecouture** [comment]
@coderabbitai review

**Comment by tylerecouture** [comment]
@coderabbitai review

File: src/courses/models.py [hunk 1]
```diff
@@ -56,15 +59,54 @@ def split_xp_between_registrations(registrations, total, xp_by_course):
...
+    So the odd XP is handed out a whole point at a time to the shares that came closest to
+    earning it (the largest-remainder method), leaving every course within 1 XP of its exact
+    share and the shares adding up to what the student actually earned. Ties go to the
+    registration made first, so a student is divided the same way however and whenever they
+    are asked about: what they are shown all term is what gets archived.
+
+    Args:
+        registrations (list[CourseStudent]): the registrations the shares belong to, in the
+            same order, for a stable tiebreak.
+        exact_shares (list[float]): each registration's exact share.
```
[the remark this answers is not in the document]
Comment (tylerecouture) [reply src/courses/models.py:92]: Fair, and the order guarantee mattered more than the tie-break: `split_xp_between_registrations()` zips the returned list straight back onto its registrations, so "in the order given" is a contract the caller depends on and the docstring never stated it. Both are now spelled out in `2fd55ffa`:

> Shares owed equally are settled by the lowest CourseStudent.pk, the registration made first: that keeps the division a property of the student rather than of the order the caller holds them in, so what a student is shown all term is what gets archived.

and in `Returns:`, "the same shares in whole XP, **in the order given**, adding up to the exact shares' total". The `Args:` entry for `registrations` now says "in the same order as exact_shares, for the tiebreak between equally owed shares" rather than the vaguer "for a stable tiebreak".

Worth noting why `pk` rather than the input order, since that is the part a future reader is most likely to want to change: the live path reaches this through `current_courses()` and the archive path through `get_semester()`, and those can hand the registrations over in different orders. Keying the tie-break on the row itself is what stops the marks page and the archived record disagreeing about which course got the odd point.

- Claude Code (`bytedeck - semester workflow redesign`)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/courses/tests/test_models.py [hunk 2]
```diff
@@ -1525,6 +1525,87 @@ def test_calc_semester_grades__records_each_course_its_own_final_xp(self):
         self.assertEqual(maths_registration.final_xp, 60)
         self.assertEqual(art_registration.final_xp, 0)
 
+    def test_calc_semester_grades__records_the_odd_xp_rather_than_dropping_it(self):
+        """final_xp is a whole number of XP, so a share that is not whole was rounded down on
+        its own and the rest of the student's work vanished from their record for good (issue
+        #2490): 5 XP shared between two courses was archived as 2 and 2. The odd XP now goes
+        to one of them, so what is recorded adds up to what the student earned."""
```
[the remark this answers is not in the document]
Comment (tylerecouture) [reply src/courses/tests/test_models.py]: Right, and this is CLAUDE.md's own rule ("comments must describe the code as it is now, never the code that was removed"), so it should not have got past me. Once this squash-merges, "was archived as 2 and 2" and "now goes to one of them" point at a diff nobody can see.

Rewritten in `2fd55ffa` as the invariant:

> Archiving records whole XP that adds up to what the student earned. final_xp is an integer field, so a share of 2.5 rounded on its own drops the odd XP from the record for good: 5 XP shared between two courses has to archive as 2 and 3, not 2 and 2 (issue #2490).

That keeps the failure mode a regression test should name (rounding each share alone loses XP) without framing it as a change from a previous implementation.

Two neighbouring docstrings had the same problem and are fixed in the same commit, since they would have aged the same way:

* `test_calc_semester_grades__hands_the_odd_xp_out_a_point_at_a_time` said "are archived as 4, 3 and 3 **rather than 3 each**", comparing against the old even split. Now: "The odd XP goes whole to the share closest to earning it, rather than being split again or dropped".
* `test_calc_semester_grades__records_the_xp_the_student_was_shown_all_term` said "**Without that**, 7 XP split two ways **would** read as 3.5". Now states what it does read: "7 XP split two ways reads as 4 and 3 on the marks page and archives as 4 and 3".

Same for the `whole_xp_shares()` docstring itself, covered in the other thread.

- Claude Code (`bytedeck - semester workflow redesign`)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by tylerecouture** [comment]
@coderabbitai review

(Requesting a pass over `4d352dde`: the two findings from the 17:18Z review are fixed in `2fd55ffa` and `develop` is merged in, but the review state is still Changes requested against the old head `27fdcbdd`, which leaves the PR blocked.)

- Claude Code (`bytedeck - semester workflow redesign`)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by tylerecouture** [comment]
@coderabbitai review

- Claude Code (`bytedeck - semester workflow redesign`)

---
_Generated by [Claude Code](https://claude.ai/code)_
